File: CT_reconstruction.m

img=imread("./image.jpg");
img_resized=imresize(img,[512,512]);
if size(img_resized,3)==3
    img_gray=rgb2gray(img_resized);
else
    img_gray=img_resized;
end
%imshow(img_gray);
% define the angle
theta_3 =0:3:179;
theta_03=0:0.3:179;
%calculate sinogram
[R_3,xp_3]=radon(img_gray,theta_3);
[R_03,xp_03]=radon(img_gray,theta_03);

%show sinogram
figure;%new window
subplot(2, 4, 1);
% left image
imshow(img_gray, []);
title('Original Image');
% right image
subplot(2, 4, 5);
imshow(img_gray, []);
title('Original Image');


subplot(2, 4, 2);
% left image
imagesc(theta_3,xp_3,R_3);
xlabel('Angle (degrees)');
ylabel('Sensor Position');
title('Sinogram 3');
colormap gray;
colorbar;
% right image
subplot(2, 4, 6);
imagesc(theta_03,xp_03,R_03);
xlabel('Angle (degrees)');
ylabel('Sensor Position');
title('Sinogram 0.3');
colormap gray;
colorbar;

% back-projection, no filter
recon_unfiltered_3 = iradon(R_3, theta_3, 'linear', 'none');  % 'none' = no filter
recon_unfiltered_03 = iradon(R_03, theta_03, 'linear', 'none');

% left image
subplot(2, 4, 3);
imshow(recon_unfiltered_3, []);
title('Sinogram 3');
% right image
subplot(2, 4, 7);
imshow(recon_unfiltered_03, []);
title('Sinogram 0.3');

% back-projection
recon_filtered_3 = iradon(R_3, theta_3, 'linear', 'Ram-Lak');  
recon_filtered_03 = iradon(R_03, theta_03, 'linear', 'Ram-Lak');

% left image
subplot(2, 4, 4);
imshow(recon_filtered_3, []);
title('Sinogram 3 filtered');
% right image
subplot(2, 4, 8);
imshow(recon_filtered_03, []);
title('Sinogram 0.3 filtered');

File: CT_reconstruction.py
import numpy as np
import matplotlib.pyplot as plt
from skimage.transform import radon, iradon
from skimage.io import imread
from skimage.color import rgb2gray
from skimage.transform import resize
# https://chatgpt.com/share/681f89d1-a3f4-8006-9272-f400f93e72a6
# 讀取並預處理影像
img = imread("C:\\Users\\User\\Desktop\\toy.jpg")
img_gray = rgb2gray(img)
img_resized = resize(img_gray, (512, 512))

# 設定角度（theta）
theta_3 = np.arange(0., 180., 3)
theta_03 = np.arange(0., 180., 0.3)

# 建立 sinogram
sinogram_3 = radon(img_resized, theta=theta_3, circle=False)
sinogram_03 = radon(img_resized, theta=theta_03, circle=False)

# 重建（Ram-Lak 濾波）
nof_reconstruction_3 = iradon(sinogram_3, theta=theta_3, filter_name=None, output_size=512, circle=False)
nof_reconstruction_03 = iradon(sinogram_03, theta=theta_03, filter_name=None, output_size=512, circle=False)

# 重建（Ram-Lak 濾波）
reconstruction_3 = iradon(sinogram_3, theta=theta_3, filter_name='ramp', output_size=512, circle=False)
reconstruction_03 = iradon(sinogram_03, theta=theta_03, filter_name='ramp', output_size=512, circle=False)

# 顯示結果
fig, axs = plt.subplots(2, 4, figsize=(15, 10))

# 第一列：每 3 度
axs[0, 0].imshow(img_resized, cmap='gray')
axs[0, 0].set_title("Original Image (3°)")

axs[0, 1].imshow(sinogram_3, cmap='gray', aspect='auto')
axs[0, 1].set_title("Sinogram (3°)")

axs[0, 2].imshow(nof_reconstruction_3, cmap='gray')
axs[0, 2].set_title("Reconstructed Image (3°) no filter")

axs[0, 3].imshow(reconstruction_3, cmap='gray')
axs[0, 3].set_title("Reconstructed Image (3°)")

# 第二列：每 0.3 度
axs[1, 0].imshow(img_resized, cmap='gray')
axs[1, 0].set_title("Original Image (0.3°)")

axs[1, 1].imshow(sinogram_03, cmap='gray', aspect='auto')
axs[1, 1].set_title("Sinogram (0.3°)")

axs[1, 2].imshow(nof_reconstruction_03, cmap='gray')
axs[1, 2].set_title("Reconstructed Image (0.3°) no filter")

axs[1, 3].imshow(reconstruction_03, cmap='gray')
axs[1, 3].set_title("Reconstructed Image (0.3°)")

# 移除座標軸
for ax in axs.flat:
    ax.axis('off')

plt.tight_layout()
plt.show()